text/0008-ingest-gem-content.md

```diff
@@ -0,0 +1,337 @@
+
+The table will be indexed to support fast lookups for common operations.
+The most common is likely "list files for a version of a gem" which will naturally index on the gem `version_id`.
+Other common operations may surface, like finding all versions of the same file within a gem, or all files with the same checksum within a gem.
+
+The table is likely to be large with hundreds of millions of rows.
+Performance must be considered to reduce its impact on the rubygems.org database as a whole.
+If performance becomes a concern, a generated json manifest can be stored along with the gem contents in S3 which would solve for the most common requests.
+
+### Recording, but not storing binary files
+
+Files that are identified as binary will be flagged in the index and not uploaded to the storage bucket.
+This is both to save storage space and avoid potential abuse.
+Large compiled binaries, image data, or other binary data is not easily browsed or diffed.
+Both npmjs.com's code browsing and github.com excludes binary files from display.
+
+### Using SHA256 to reduce storage size and increase cache hits
+
+As of this writing, there is more than 5TB of gem files stored by rubygems.org.
+
+Using `rake` as an example, the total size of all `.gem` files is currently 6.8MB.
+
+The average compression across the entire history of the rake gem is about 84%.
+Expanded fully, all versions of rake add up to 42MB.
+
+Many files in rake do not change between versions, allowing us to optimize the storage.
+If each file is instead saved at its SHA256 checksum, causing and identical file to be stored only once, the size of all uncompressed files in `rake` droms from 42MB to 10MB.
+The resulting directory has 1226 files for all versions of `rake`.
+
+In order to access a hashed filename, access the file content table to find all files in the version of rake.
+Use the relative file path within the gem to find the SHA256 checksum.
+Access the S3 bucket at a key similar to `/rake/files/c588eda0547f6fa72fd7cd9ba2ddc2e81b96e61d690a929e8f579412da27eebd`
+
+### Yanking gems
+
+The hashed filenames pose a slight challenge when yanking a gem.
+The relative infrequency of yanking means that we can accept a more involved process removing files than adding them or hosting them.
+
+When a version is yanked, the file manifest for that version will be used to search for any hashed files that are unique to the yanked gem.
+If a SHA256 named file would be orphaned by yanking the gem (it has no gem versions that reference it) then the file will be deleted from S3 and purged from Fastly's cache.
+If a file has other references, then the file was already a duplicate of a file in a previous version. Files with multiple references will be preserved until the last reference is removed.
+
+When a gem is yanked the manifest table rows recording the file names with checksum, size could be preserved but no longer shown anywhere.
```

Comment:
Should there be a "status" or "yanked T/F?" column in the table where we can deterministically record if a file was yanked? (Or does Ruby have a nice "soft delete" feature built in we can coopt for this purpose?) 

Since "a record with no source file" would under normal operations be considered a "bad" or "error" state, I think it's worth tracking when we MEANT to get something into that state so we can differentiate between actual errors and intentional actions.

Reply:
Ah! I see you had a similar thought below. :D

Reply:
I think we'd be able to check the yank status by looking at the referenced gem version row